### PR TITLE
Enhance person view

### LINKED ISSUE #3235: Enhance Person View layout
Currently, the person view page is not very user friendly. It is long and doesn't look nice when printed. The order of the fields can't be changed.

It would be great to have an improved Person view with the following features:

- Responsive layout ( for small screens and preview areas in addition to standard screen sizes)
- fields (both custom and regular) to be ordered as per order defined in `anet-dictionary.yml`. This will require merging `customFields` with the parent.

in the case of normal screen we can have:

<table>
<tbody>
  <tr>
    <td rowspan="3" style=”font-family: monospace”>///\<br>||oo<br>|| &gt;<br> \_-</td>
    <td><span style="font-style:italic">field2</span>: FIELD 2</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field3</span>: FIELD 3</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field4</span>: FIELD 4</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field1</span>: FIELD 1</td>
    <td><span style="font-style:italic">field5</span>: FIELD 5</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field6</span>: FIELD 6</td>
    <td><span style="font-style:italic">field7</span>: FIELD 7</td>
  </tr>
</tbody>
</table>

(alternatively, we could have the avatar on the right, and sequential position of fields. This might be a better solution) 

in the case of small screen we can have:

<table>
<tbody>
  <tr>
    <td style=”font-family: monospace”>///\<br>||oo<br>|| &gt;<br> \_-</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field1</span>: FIELD 1</td>
  </tr>
  <tr>
    <td><span style="font-style:italic">field2</span>: FIELD 2</td>
  </tr>
</tbody>
</table>

### PR DESCRIPTION
**Pull request by cemalettin-simsoft**
Person page updated.

#### User changes
- Users will see a more compact Person page, more responsive.

#### Super User changes
-

#### Admin changes
- Admins can add/remove/order the person fields show in the Person page.

#### System admin changes
-
- [X] anet.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [ ] Described the user behavior in PR body
  - [ ] Referenced/updated all related issues
  - [ ] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [ ] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [ ] Added and/or updated unit tests
  - [ ] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [ ] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

File: client/src/periodUtils.js [hunk 1]
```diff
@@ -159,3 +159,41 @@ PeriodsTableHeader.propTypes = {
...
+
+function getPeriodNumberForScreen(width) {
+  for (const size in SCREEN_SIZE_TO_PERIOD_NUMBER) {
+    if (SCREEN_SIZE_TO_PERIOD_NUMBER[size].isMatch(width)) {
+      return SCREEN_SIZE_TO_PERIOD_NUMBER[size].no
+    }
+  }
+}
+
+export const useLessNumberOfPeriodsOnSmallScreens = setNumberOfPeriods => {
```
Comment (gjvoosten) [inline client/src/periodUtils.js:194]: I'd rename this function to `useFewerPeriodsOnSmallScreens` ("less periods" is grammatically incorrect, and on top of that it looks like a "useless" function :wink: ).

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/periodUtils.js (hunk 1 above)
Comment (cemalettin-simsoft) [reply client/src/periodUtils.js:194]: Ah, yes. Actually, i should probably rename something like "useResponsiveNumberOfPeriods". Hiding the "Desktop-first" bias :smile:

File: anet-dictionary.yml [hunk 2]
```diff
@@ -321,18 +321,34 @@ dictionary:
...
-
+    # cF_ => lifted custom fields
     person:
-      firstName: First name
-      lastName: Last name
-      domainUsername: Domain username
+      numberOfFieldsUnderAvatar: 7
+      firstName:
+        label: First name
+        inShowPage: false
```
Comment (gjvoosten) [inline anet-dictionary.yml:329]: Wouldn't it be easier to just specify a list of fields show on the Show page, e.g.:
```
      fieldsOnShowPage:
      - domainUsername
      - emailAddress
      - …etc…
```
and leave the rest unchanged?

File: anet-dictionary.yml [hunk 3]
```diff
@@ -640,7 +662,7 @@ dictionary:
 
       position:
         type: ANET Administrator
-
+  liftedCustomFieldPrefix: cF_
```
Comment (gjvoosten) [inline anet-dictionary.yml:665]: Wouldn't it be easier to just specify a list of 'lifted custom fields', e.g.:
```
      liftedCustomFields:
      - multipleButtons
      - inputFieldName
      - …etc…
```
?

File: src/main/resources/anet-schema.yml [hunk 4]
```diff
@@ -461,56 +461,38 @@ properties:
 
       person:
         type: object
-        additionalProperties: false
```
Comment (gjvoosten) [inline src/main/resources/anet-schema.yml:464]: Why did you remove this? It was there so admins don't add non-supported properties wihout being warned about this.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/pages/people/Show.js [hunk 5]
```diff
@@ -174,6 +183,15 @@ const PersonShow = ({ pageDispatchers }) => {
           <a href={`mailto:${person.emailAddress}`}>{person.emailAddress}</a>
         )
 
+        const orderedFields = orderPersonFields()
+        const numberOfFieldsUnderAvatar =
+          Settings.fields.person.numberOfFieldsUnderAvatar || 3
+        const leftColumUnderAvatar = orderedFields.slice(
+          0,
+          numberOfFieldsUnderAvatar - 1
```
Comment (gjvoosten) [inline client/src/pages/people/Show.js:191]: `slice` excludes the end index, so that should be:
```
          numberOfFieldsUnderAvatar
```
You're missing a field now (and the number of fields under the avatar is 6 when the dictionary wants 7).

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: anet-dictionary.yml (hunk 2 above)
Comment (cemalettin-simsoft) [reply anet-dictionary.yml:329]: My first thinking was doing like this, then i thought, this would duplicate the field names, needs to be synced whenever something is added etc.

File: client/src/pages/people/Show.js [hunk 6]
```diff
@@ -403,15 +364,7 @@ const PersonShow = ({ pageDispatchers }) => {
...
-
-              {Settings.fields.person.customFields && (
-                <Fieldset title="Person information" id="custom-fields">
-                  <ReadonlyCustomFields
-                    fieldsConfig={Settings.fields.person.customFields}
-                    values={values}
-                  />
-                </Fieldset>
-              )}
+              )
```
Comment (gjvoosten) [inline client/src/pages/people/Show.js:367]: Spurious closing parenthesis.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: anet-dictionary.yml (hunk 2 above)
Comment (gjvoosten) [reply anet-dictionary.yml:329]: From my point of view, having these together in one list is clearer and more convenient.

File: src/main/resources/anet-schema.yml (hunk 4 above)
Comment (cemalettin-simsoft) [reply src/main/resources/anet-schema.yml:464]: I did remove because custom fields were lifted, so there were additional properties

File: anet-dictionary.yml (hunk 3 above)
Comment (cemalettin-simsoft) [reply anet-dictionary.yml:665]: We want to order normal fields and custom fields mixed up. That is why i need them in the same block

File: client/src/components/CustomFields.js [hunk 7]
```diff
@@ -936,6 +939,49 @@ ReadonlyCustomFields.defaultProps = {
...
+export const mapReadonlyCustomFieldsToComps = ({
+  fieldsConfig,
+  parentFieldName, // key path in the values object to get to the level of fields given by the fieldsConfig
+  values,
+  vertical
+}) => {
+  return Object.keys(fieldsConfig).reduce((accum, key) => {
+    const fieldName = `${parentFieldName}.${key}`
+    const fieldConfig = fieldsConfig[key]
+    const fieldProps = getFieldPropsFromFieldConfig(fieldConfig)
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/components/CustomFields.js]: Yes, there are a lot of similar code in the project when i search `Object.keys`. For now only changed the `CustomFields.js` file

File: anet-dictionary.yml (hunk 3 above)
Comment (gjvoosten) [reply anet-dictionary.yml:665]: Then why not combine the two lists (`fieldsOnShowPage` and `liftedCustomFields`) into one list to specify which fields (standard & custom) need to be shown in which order on the Show page?
Just having this extra list prop makes field definitions the same again for all object types, and adding similar functionality to other Show pages would involve just adding such a list to an object type (and adapting its Show page to use it, obviously). And then it makes sense to rename `numberOfFieldsUnderAvatar` to something like `numberOfFieldsIn FirstColumn`.

File: src/main/resources/anet-schema.yml (hunk 4 above)
Comment (gjvoosten) [reply src/main/resources/anet-schema.yml:464]: Another reason not to do it like this, but to do something like https://github.com/NCI-Agency/anet/pull/3278#discussion_r517923184

File: anet-dictionary.yml (hunk 3 above)
Comment (cemalettin-simsoft) [reply anet-dictionary.yml:665]: Ah yes we are getting to the critical part. I think i wanted to separate the custom and non-custom ones, because there was already a customField parser component. `ReadonlyCustomFields`. Because i had the info of non-custom component fields and props in show page, I can map each non-custom component like this:
```
          return Object.keys(Person.shownNonCustomFields).reduce(
            (accum, key) => {
              accum[key] = (
                <Field
                  name={key}
                  label={Person.shownNonCustomFields[key].label}
                  component={FieldHelper.ReadonlyField}
                  humanValue={humanValuesExceptions[key]}
                  className={classNameExceptions[key]}
                />
              )

              return accum
            },
            {}
          ) 
```
I can't map custom fields like this, because there are configs to go through, and i wanted to reuse the already built-in config parsing file `CustomFields.js`

File: client/src/index.css [hunk 8]
```diff
@@ -1043,57 +1071,126 @@ header.searchPagination {
...
+  /*	Fixes the print style for the main viewport. Now the entire page is printed instead of the 100vh
+    This resets the display to block and height to 100%. And overrides the viewport height and flexbox display. */
+  #main-viewport,
+  .anet,
+  .content-container {
+    display: block !important;
+    height: 100% !important;
+  }
+}
+
```
Comment (cemalettin-simsoft) [inline client/src/index.css:1148]: Non stylistic changes start from here

**Comment by cemalettin-simsoft** [comment]
This branch is merged into https://github.com/NCI-Agency/anet/pull/3298, will continue there.
